js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,323 @@
+
+        // Update time
+        let timestamp = params.timestamp;
+        let time = timestamp ? GLib.DateTime.new_from_unix_local(timestamp) : null;
+
+        this.update('', params.body.unpack(), {gicon: newGicon, datetime: time});
+    },
+});
+
+var ClubhouseNotificationSource = new Lang.Class({
+    Name: 'ClubhouseNotificationSource',
+    Extends: NotificationDaemon.GtkNotificationDaemonAppSource,
+
+    _init: function(appId) {
+        this.parent(appId);
+        this._notification = null;
+    },
+
+    addNotification: function(notificationId, notificationParams, showBanner) {
+        if (this._notification == null)
+            this._notification = new ClubhouseNotification(this, notificationParams);
+        else
+            this._notification.updateFromParams(notificationParams);
+
+        this.notify(this._notification);
+    },
+});
+
+var ClubhouseComponent = new Lang.Class({
+    Name: 'ClubhouseComponent',
+
+    _init: function() {
+        this._useClubhouse = false;
+
+        let eosImageVersion = Shell.util_get_eos_image_version();
+        if (eosImageVersion == null || !this._imageUsesClubhouse(eosImageVersion))
+            return;
+
+        this._useClubhouse = true;
+        this._banner = null;
+        this.actor = null;
+        this.source = null;
+        this._clubhouseProxy = null;
+        this._clubhouseProxyHandler = 0;
+        this._oldAddNotificationFuncPrototype = null;
+    },
+
+    enable: function() {
+        if (!this._useClubhouse) {
+            log('Cannot enable Clubhouse in this image version');
+            return;
+        }
+
+        log('Enabling Clubhouse component');
+
+        this._clubhouseProxy = new Gio.DBusProxy.new_sync(sessionBus,
+                                                          Gio.DBusProxyFlags.DO_NOT_AUTO_START_AT_CONSTRUCTION,
+                                                          null,
+                                                          CLUBHOUSE_ID,
+                                                          CLUBHOUSE_DBUS_OBJ_PATH,
+                                                          CLUBHOUSE_ID,
+                                                          null);
+        this._clubhouseProxy.connect('notify::g-name-owner', Lang.bind(this, function() {
+            if (!this._clubhouseProxy.g_name_owner) {
+                log('Nothing owning D-Bus name %s, so dismiss the Clubhouse banner'.format(CLUBHOUSE_ID));
+                this._dismiss_quest_cb();
+            }
+        }));
+
+        this.actor = new St.Widget({ visible: true,
+                                     clip_to_allocation: true,
+                                     can_focus: true,
+                                     track_hover: true,
+                                     reactive: true,
+                                     layout_manager: new Clutter.BinLayout() });
+
+        Main.layoutManager.addChrome(this.actor);
```

Comment:
Not sure if you saw this question.

Reply:
Sorry, I did see it but forgot to reply.
I copied this from the old Game Manager I believe. You seemed to me as well that this would give us a better control if, e.g. we decide to show two banners. But if you want to change it, please go ahead (I can no longer stare at the screen today).